### ISSUE TITLE
Rewards based on predicted Action

### ISSUE THREAD

**Issue by CodingWorld14** [m4saka/xxr#17]
Hello,

Say I have the following:
```
// Available actions in dataset
std::unordered_set<int> availableActions{ 0, 1, 2, 3, 4, 5};
```
And I want to add 100 points if the action '2' is chosen, subtract 50 points if the action '3' is chosen, and do nothing for the others.

A simple reward sum can be shown below:
```
// Test
    double rewardSum = 0.0;
    for (int i = 0; i < 300000; ++i)
    {
        // Get current situation from environment
        std::vector<int> situation = env.situation();

        // Select action by XCS
        int action = xcs.exploit(situation);

        // Execute action and receive reward from environment
        rewardSum += env.executeAction(action);
    }
```

Is this possible to do? And would this be beneficial?

**Comment by m4saka** [comment]
You can do it by creating your own environment (problem) class.

Using different rewards for different classes won't cause any problems for XCS to work, but doing it won't change any results.
What is your true purpose in doing it?

**Comment by m4saka** [comment]
Maybe you mean that you want to change the performance displayed after the evaluation, rather than the reward during training?
If so, you just need to do the following:

```cpp
// Test
double rewardSum = 0.0;
for (int i = 0; i < 300000; ++i)
{
    // Get current situation from environment
    std::vector<int> situation = env.situation();

    // Select action by XCS
    int action = xcs.exploit(situation);

    // Execute action and receive reward from environment
    double reward = env.executeAction(action);
    if (reward > 0.0)
    {
        switch (action)
        {
        case 2:
            rewardSum += reward + 100.0;
            break;
        case 3:
            rewardSum += reward - 50.0;
            break;
        default:
            rewardSum += reward;
        }
    }
}
```

**Comment by CodingWorld14** [comment]
Thanks for your response.

What does the feeding back the reward to XCS do?
```
// Feedback the reward to XCS
xcs.reward(reward);
```

I guess I am working out the best way to go about building out my reward function. Ultimately, I want the XCS model to choose the correct output based on its inputs, that is the primary goal. But I also want some outputs to carry a greater reward than others, not necessarily a better choice, but that it was a "better" answer.

That may be a poor explanation, and it kind of goes back to it depends what that ```xcs.reward(reward);``` does.

Is there a feedback into the model, so that the XCS is learning over time and being changed by the reward? (Ex. if you have a model that is training on 100k iterations, the reward is changing the models learning as it is going?)

Ex.
Inputs:                                      Predicted Output:
0  1   2                                      3
0  2   1                                      4
0  1   2                                      5
0  2   1                                      4

In the first row, we have inputs (0 1 2), and the XCS predicts 3, which we know is wrong, but the XCS doesn't know that 'yet'. So we feedback in our reward system that the inputs (0 1 2) along with the predicted output of 3 is **bad** and we assign it a reward of -50.

The next row, we have inputs (0 2 1) and the XCS predicts 4, which is correct, so we give the inputs (0 2 1) and the out of 4 a thumbs up, and give it a reward of 100 for doing something we liked.

The third row, we again have the inputs (0 1 2) and the XCS this time predicts 5, which is the actual correct answer that we are looking for. We feedback into our reward system that the inputs (0 1 2) along with the predicted output of 5 is **good** and we give it a reward of 100. Now we know based on our feedback that the inputs (0 1 2) is penalized for predicting 3, and rewarded for predicting 5.

The fourth row goes again, and because in the second row we were rewarding based on our answer with the same inputs,  we will tend to favor that same predicted output, so we say 4 again.

Does this kind of make sense?

**Comment by m4saka** [comment]
You don't need to do that because `env.executeAction(action);` is already designed to return "1000.0" if the answer prediction (`action`) is correct; otherwise, it returns "0.0".

**Comment by CodingWorld14** [comment]
How does the answer prediction ``` ( action ) ``` know that it is correct without some type of outside reward system? If I feed 1000 lines of data into XCS, it will be able to create classifiers and their associated experience and action etc... but without me explicitly telling it what the correct action is, how would it know?

**Comment by m4saka** [comment]
The environment class (like `DatasetEnvironment`) is what you call an "external reward system".
XCS receives a state from the environment class and determines an action (output), and the environment class returns a reward value of 0 or 1000 depending on whether it matches the answer.

**Comment by CodingWorld14** [comment]
Gotcha, that makes sense. Would there ever be a benefit to returning a higher reward to specific actions (output)?

Similar to the ``` switch (action) ``` statement you posted above, and put that inside of  ``` executeAction ``` .

```
virtual double executeAction(Action action) override
        {
            const double reward = (action == m_answer) ? 1000.0 : 0.0;

            // Single-step problem
            m_isEndOfProblem = true;

            loadNext();

            return reward;
        }
```

**Comment by m4saka** [comment]
No, using different rewards for different classes does not change the results at all.
This is because the XCS does not employ the rewards themselves as the fitness of the classifier.

XCS is often referred to as an "accuracy-based learning classifier system". Here, "accuracy" means the smallness of the error in reward predictions.
XCS generates a different classifier for each different reward, and selects the classifiers whose reward predictions are "accurate".
This does NOT mean that XCS selects the classifier having high reward values.
(For example, a classifier that accurately predicts a reward of "0" will be selected equally well as that of "1000".)

Let's say you have two types of rewards, 1000 and 0.
In that case, two classifiers are generated for a state (input) and behavior (output) pair: one predicts a reward value of 1000 and another predicts a reward value of 0.
In other words, XCS generates classifiers that predicts both correct and incorrect answer for an input in classification problems.
In this case, even if you change the reward values to "800 and 0" or to "100 and -50" for particular classes, the number of classifiers to be generated is the same, so it has no effect on the learning.
(Not exactly the same, but it won't have the effect you're hoping for. In order to get the same results using different scales for rewards, you need to change a parameter called ε0. This is because ε0 defines the tolerance of the reward that can be considered the same reward value.)

In conclusion, there is no advantage to using reward values other than 0 and 1000 when solving classification problems in XCS.

**Comment by CodingWorld14** [comment]
Thanks so much for your reply and well thought out answer! This helps clarify my understanding a great deal!

**Comment by m4saka** [comment]
ε_0 is in `XCSConstants`.

You can set the value by the following change:
```diff
 int main()
{
     // XCS hyperparameters
     xxr::XCSConstants constants;
     constants.n = 1200; // N (max number of classifiers)
+    constants.epsilonZero = 10.0; // epsilonZero (error threshold of reward prediction, default:10)
```
1% of the result value range is recommended; you should use 10.0 for the reward 0–1000, and 1.0 for the reward 0–100.
